privacyidea/lib/token.py

```diff
@@ -538,9 +541,24 @@ def get_tokens_paginate(tokentype=None, realm=None, assigned=None, user=None,
            "next": next,
            "current": page,
            "count": pagination.total}
+    if counts is True:
+        ret.update({"counts": get_token_type_counts(sql_query)})
     return ret
 
 
+def get_token_type_counts(token_sql_query_obj):
+    """
+    This function extracts the numbers of tokens per tokentype from the given sql query object
+    """
+    token_list_full = token_sql_query_obj.all()
+    tokentypes_in_list = set([token.tokentype for token in token_list_full])
+    counts = dict.fromkeys(tokentypes_in_list, "")
+    for tokentype in tokentypes_in_list:
```

Comment:
Do not do it this way.
You do not want to fetch 50.000 tokens from the database and iterate through them!
This will be getting slow.

This way you are issuing a second sql select statement on all tokens.
This first statement will only fetch the pagination, but this will fetch all tokens.

Reply:
Sadly, you are right. Let's take a step back an think again.